internal/resource/mappers/bimapper/bimapper.go

```diff
@@ -0,0 +1,207 @@
+// The primary object is called the "item" and an item has many "links".
+// Tracking is done on items.
+type Mapper struct {
+	itemType, linkType *pbresource.Type
+
+	lock       sync.Mutex
+	itemToLink map[resource.ReferenceKey]map[resource.ReferenceKey]struct{}
+	linkToItem map[resource.ReferenceKey]map[resource.ReferenceKey]struct{}
+}
+
+// New creates a bimapper between the two required provided types.
+func New(itemType, linkType *pbresource.Type) *Mapper {
+	if itemType == nil {
+		panic("itemType is required")
+	}
+	if linkType == nil {
+		panic("linkType is required")
+	}
+	return &Mapper{
+		itemType:   itemType,
+		linkType:   linkType,
+		itemToLink: make(map[resource.ReferenceKey]map[resource.ReferenceKey]struct{}),
+		linkToItem: make(map[resource.ReferenceKey]map[resource.ReferenceKey]struct{}),
+	}
+}
+
+// Reset clears the internal mappings.
+func (m *Mapper) Reset() {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	m.itemToLink = make(map[resource.ReferenceKey]map[resource.ReferenceKey]struct{})
+	m.linkToItem = make(map[resource.ReferenceKey]map[resource.ReferenceKey]struct{})
+}
+
+// IsEmpty returns true if the internal structures are empty.
+func (m *Mapper) IsEmpty() bool {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	return len(m.itemToLink) == 0 && len(m.linkToItem) == 0
+}
+
+// UntrackItem removes tracking for the provided item. The item type MUST match
+// the type configured for the item.
+func (m *Mapper) UntrackItem(item *pbresource.ID) {
+	if !resource.EqualType(item.Type, m.itemType) {
+		panic(fmt.Sprintf("expected item type %q got %q",
+			resource.TypeToString(m.itemType),
+			resource.TypeToString(item.Type),
+		))
+	}
+	m.untrackItem(resource.NewReferenceKey(item))
+}
+
+func (m *Mapper) untrackItem(item resource.ReferenceKey) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	m.removeItemLocked(item)
+}
+
+// TrackItem adds tracking for the provided item. The item and link types MUST
+// match the types configured for the items and links.
+func (m *Mapper) TrackItem(item *pbresource.ID, links []*pbresource.Reference) {
+	if !resource.EqualType(item.Type, m.itemType) {
+		panic(fmt.Sprintf("expected item type %q got %q",
+			resource.TypeToString(m.itemType),
+			resource.TypeToString(item.Type),
+		))
+	}
+
+	linksAsKeys := make([]resource.ReferenceKey, 0, len(links))
+	for _, link := range links {
+		if !resource.EqualType(link.Type, m.linkType) {
+			panic(fmt.Sprintf("expected link type %q got %q",
+				resource.TypeToString(m.linkType),
+				resource.TypeToString(link.Type),
+			))
+		}
+		linksAsKeys = append(linksAsKeys, resource.NewReferenceKey(link))
+	}
+
+	m.trackItem(resource.NewReferenceKey(item), linksAsKeys)
+}
+
+func (m *Mapper) trackItem(item resource.ReferenceKey, links []resource.ReferenceKey) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	m.removeItemLocked(item)
+	m.addItemLocked(item, links)
+}
+
+func (m *Mapper) removeItemLocked(item resource.ReferenceKey) {
+	for link := range m.itemToLink[item] {
```

Comment:
does this need to lock and unlock?

Reply:
My naming convention is to end a helper function with `xxxLocked` if the method is to be called while holding the lock, as in the preceding function `trackItem`. I can also decorate all of these functions with lock usage commentary.

Reply:
I see. I had interpreted this to mean that this function will do the locking. But perhaps I'm mistaking the convention here.